org.eclipse.scanning.example/src/org/eclipse/scanning/example/detector/PosDetector.java

```diff
@@ -0,0 +1,161 @@
+			final int uniqueKey = position.getStepIndex() + 1;
+			final Dataset uniqueKeyDataset = DatasetFactory.createFromObject(uniqueKey);
+			rslice = IScanRankService.getScanRankService().createScanSlice(position);
+			sliceND = new SliceND(uniqueKeys.getShape(), uniqueKeys.getMaxShape(),
+					rslice.getStart(), rslice.getStop(), rslice.getStep());
+			uniqueKeys.setSlice(null, uniqueKeyDataset, sliceND);
+			
+		} catch (DatasetException e) {
+			setDeviceState(DeviceState.FAULT);
+			throw new ScanningException("Failed to write the data to the NeXus file", e);
+		}
+		
+		setDeviceState(DeviceState.ARMED);
+		return true;
+	}
+	
+	private void createNexusFile() throws ScanningException {
+		TreeFile treeFile = NexusNodeFactory.createTreeFile(filePath);
+		NXroot root = NexusNodeFactory.createNXroot();
+		treeFile.setGroupNode(root);
+		NXentry entry = NexusNodeFactory.createNXentry();
+		root.setEntry(entry);
+		
+		NXcollection ndAttributesCollection = NexusNodeFactory.createNXcollection();
+		entry.setCollection("NDAttributes", ndAttributesCollection);
+		uniqueKeys = ndAttributesCollection.initializeLazyDataset("NDArrayUniqueId", scanRank, Integer.class);
+		
+		NXdata dataGroup = NexusNodeFactory.createNXdata();
+		entry.setData(getName(), dataGroup);
+		data = dataGroup.initializeLazyDataset(NXdata.NX_DATA, scanRank + 2, Double.class);
+		
+		INexusFileFactory nff = ServiceHolder.getNexusFileFactory();
+		NexusFile file = nff.newNexusFile(filePath, true);
+		try {
+			file.createAndOpenToWrite();
+			file.addNode("/", treeFile.getGroupNode());
+			file.activateSwmrMode();
+			file.flush();
+		} catch (NexusException e) {
+			throw new ScanningException(e);
+		}
+	}
+	
+	@Override
+	public NexusObjectProvider<NXdetector> getNexusProvider(NexusScanInfo info) throws NexusException {
+		scanRank = info.getRank();
+		filePath = getFilePath(info);
+				
+		NXdetector detector = NexusNodeFactory.createNXdetector();
+		detector.setCount_timeScalar(model.getExposureTime());
+		
+		NexusObjectWrapper<NXdetector> nexusWrapper = 
+				new NexusObjectWrapper<NXdetector>(getName(), detector, NXdetector.NX_DATA);
+		nexusWrapper.addExternalLink(detector, NXdetector.NX_DATA, getFilePath(info),
+				"/entry/" + getName() + "/" + NXdata.NX_DATA, info.getRank() + 2);
+		nexusWrapper.setPropertyValue(PROPERTY_NAME_UNIQUE_KEYS_PATH,
+				"/entry/NDAttributes/NDArrayUniqueId");
+		
+		return nexusWrapper;
+	}
+	
+	private String getFilePath(NexusScanInfo info) {
+		final File scanFile = new File(info.getFilePath());
+		final File scanDir = scanFile.getParentFile();
+		String scanFileNameNoExtn = scanFile.getName();
+		final int dotIndex = scanFileNameNoExtn.indexOf('.');
+		if (dotIndex != -1) {
+			scanFileNameNoExtn = scanFileNameNoExtn.substring(0, dotIndex);
+		}
+		final File outputDir = new File(scanDir, scanFileNameNoExtn);
+		outputDir.mkdir();
+		
+		final String filePath = new File(outputDir, "posdetector.h5").getAbsolutePath();
```

Comment:
![MAJOR](https://sonarsource.github.io/sonar-github/severity-major.png 'Severity: MAJOR') Rename "filePath" which hides the field declared at line 47. [![rule](https://sonarsource.github.io/sonar-github/rule.png)](https://sonarcloud.io/coding_rules#rule_key=squid%3AHiddenFieldCheck)
![MINOR](https://sonarsource.github.io/sonar-github/severity-minor.png 'Severity: MINOR') Immediately return this expression instead of assigning it to the temporary variable "filePath". [![rule](https://sonarsource.github.io/sonar-github/rule.png)](https://sonarcloud.io/coding_rules#rule_key=squid%3AS1488)